kafka-streams-samples/kafka-streams-recoverable/src/test/java/kafka/streams/record/recoverable/KafkaStreamsRecoverableSampleTest.java

```diff
@@ -0,0 +1,115 @@
+public class KafkaStreamsRecoverableSampleTest {
+
+    @ClassRule
+    public static EmbeddedKafkaRule embeddedKafkaRule = new EmbeddedKafkaRule(1, true,
+            "topic-in", "default-out", "custom-out");
+
+    private static EmbeddedKafkaBroker embeddedKafka = embeddedKafkaRule.getEmbeddedKafka();
+
+    private static Consumer<String, String> consumer;
+    private static Consumer<String, String> defaultOutConsumer;
+    private static Consumer<String, String> customOutConsumer;
+
+    @BeforeClass
+    public static void setUp() {
+        Map<String, Object> consumerProps = KafkaTestUtils.consumerProps("group", "false", embeddedKafka);
+        consumerProps.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        DefaultKafkaConsumerFactory<String, String> cf = new DefaultKafkaConsumerFactory<>(consumerProps);
+        consumer = cf.createConsumer();
+        embeddedKafka.consumeFromEmbeddedTopics(consumer, "topic-in", "default-out", "custom-out");
+        System.setProperty("spring.cloud.stream.kafka.streams.binder.brokers", embeddedKafka.getBrokersAsString());
+    }
+
+    @AfterClass
+    public static void tearDown() {
+        consumer.close();
+        System.clearProperty("spring.cloud.stream.kafka.streams.binder.brokers");
+    }
+
+    @Test
+    public void testKafkaStreams() {
+        Map<String, Object> senderProps = KafkaTestUtils.producerProps(embeddedKafka);
+        DefaultKafkaProducerFactory<Integer, String> pf = new DefaultKafkaProducerFactory<>(senderProps);
+        try {
+            KafkaTemplate<Integer, String> template = new KafkaTemplate<>(pf, true);
+            template.setDefaultTopic("topic-in");
+            template.send(MessageBuilder.withPayload("1").build());
+            template.send(MessageBuilder.withPayload("2").build());
+            template.send(MessageBuilder.withPayload("3").build());
+            template.send(MessageBuilder.withPayload("4").build());
+            template.send(MessageBuilder.withPayload("5").build());
+            ConsumerRecords<String, String> consumerRecords = KafkaTestUtils.getRecords(consumer);
+            var topicIn = new ArrayList<ConsumerRecord<String, String>>();
+            var defaultOut = new ArrayList<ConsumerRecord<String, String>>();
+            var customOut = new ArrayList<ConsumerRecord<String, String>>();
+
+            for(var record : consumerRecords.records("topic-in")) {
+                topicIn.add(record);
+            }
+            for(var record : consumerRecords.records("default-out")) {
+                defaultOut.add(record);
+            }
+            for(var record : consumerRecords.records("custom-out")) {
+                customOut.add(record);
+            }
+
+            assertEquals(6, topicIn.size());
+            assertThat(defaultOut.size()).isLessThanOrEqualTo(2);
+            assertThat(customOut.size()).isLessThanOrEqualTo(2);
+
+            assertEquals(0, Integer.parseInt(topicIn.get(0).value()));
```

Comment:
Where do we send record `0`?

Reply:
Removed by making the test more deterministic, there was too much variation in the first iteration of the test. This was passing but not actually correct.